cpp/src/arrow/array/array_dict.cc

```diff
@@ -212,6 +212,56 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
   return out_data;
 }
 
+struct CountDictionaryNullValuesVistor {
+  const std::shared_ptr<ArrayData>& data;
+  int64_t& out_null_count;
+
+  template <typename IndexArrowType>
+  Status CountDictionaryNullValuesImpl() {
+    int64_t index_length = data->length;
+    int64_t dict_length = data->dictionary->length;
+    const uint8_t* dictionary_null_bit_map = data->dictionary->GetValues<uint8_t>(0);
+
+    using CType = typename IndexArrowType::c_type;
+    const CType* indices_data = data->GetValues<CType>(1);
+    CType dict_len = static_cast<CType>(dict_length);
+    for (int64_t i = 0; i < index_length; i++) {
+      if (data->IsNull(i)) {
+        continue;
+      }
+
+      CType current_index = indices_data[i];
+      if (current_index < 0 || current_index >= dict_len) {
+        return Status::IndexError(
+            "Index out of bounds while counting dictionary array: ", current_index,
+            "(dictionary is ", dict_length, " long) at position ", i);
+      }
+      if (!bit_util::GetBit(dictionary_null_bit_map, current_index)) {
+        out_null_count++;
+      }
+    }
+    return Status::OK();
+  }
+
+  template <typename Type>
+  enable_if_integer<Type, Status> Visit(const Type&) {
+    return CountDictionaryNullValuesImpl<Type>();
+  }
+
+  Status Visit(const DataType& type) {
+    return Status::TypeError("Expected an Index Type of Int or UInt");
```

Comment:
How about showing the actual type?

```suggestion
    return Status::TypeError("Expected an Index Type of Int or UInt: ", type);
```